src/drivers/net/marvell/aqc1xx.c

```diff
@@ -0,0 +1,634 @@
+	tx->dx_type = 0x1;
+	tx->eop = 0x1;
+	tx->cmd = 0x22;
+	wmb();
+
+	DBG("AQUANTIA: %p TX[%d] is [%llx, %llx]\n", nic, nic->tx_ring.sw_tail,
+	    ((unsigned long long) address),
+	    ((unsigned long long) address + len));
+
+	atl_ring_next_dx(&nic->tx_ring.sw_tail);
+	ATL_WRITE_REG(nic->tx_ring.sw_tail, ATL_RING_TAIL);
+
+	return 0;
+}
+
+void atl_check_link(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+
+	uint32_t link_state = nic->hw_ops->get_link(nic);
+
+	if (link_state != nic->link_state) {
+		if (link_state) {
+			DBG("AQUANTIA: link up\n");
+			netdev_link_up(netdev);
+		} else {
+			DBG("AQUANTIA: link lost\n");
+			netdev_link_down(netdev);
+		}
+		nic->link_state = link_state;
+	}
+}
+
+/**
+* Poll for completed packets
+*
+* @v netdev		Network device
+*/
+void atl_poll_tx(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+	struct atl_desc_tx_wb *tx;
+
+	/* Check for completed packets */
+	while (nic->tx_ring.sw_head != nic->tx_ring.sw_tail) {
+
+		/* Get next transmit descriptor */
+		tx = (struct atl_desc_tx_wb *)nic->tx_ring.ring +
+		nic->tx_ring.sw_head;
+		/* Stop if descriptor is still in use */
+		if (!tx->dd)
+			return;
+
+		DBG("AQUANTIA %p: TX[%d] complete\n",
+		nic, nic->tx_ring.sw_head);
+
+		/* Complete TX descriptor */
+		atl_ring_next_dx(&nic->tx_ring.sw_head);
+		netdev_tx_complete_next(netdev);
+	}
+}
+
+/**
+* Poll for received packets
+*
+* @v netdev		Network device
+*/
+void atl_poll_rx(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+	struct atl_desc_rx_wb *rx;
+	struct io_buffer *iobuf;
+	size_t len;
+
+	/* Check for received packets */
+	while (nic->rx_ring.sw_head != nic->rx_ring.sw_tail) {
+
+		/* Get next receive descriptor */
+		rx = (struct atl_desc_rx_wb *)nic->rx_ring.ring +
+		nic->rx_ring.sw_head;
+
+		/* Stop if descriptor is still in use */
+		DBG("AQUANTIA: rx poll: desc: %llx, %llx\n", *((uint64_t *)rx),
+		    *(((uint64_t *)rx) + 1));
+		if (!rx->dd)
+			return;
+
+		/* Populate I/O buffer */
+		iobuf = nic->iobufs[nic->rx_ring.sw_head];
+		nic->iobufs[nic->rx_ring.sw_head] = NULL;
+		len = le16_to_cpu(rx->pkt_len);
+		iob_put(iobuf, len);
+
+		/* Hand off to network stack */
+		/*to do: process error*/
+		DBG("AQUANTIA: %p RX[%d] complete (length %zd)\n",
+		    nic, nic->rx_ring.sw_head, len);
+		netdev_rx(netdev, iobuf);
+		DBG("AQUANTIA %p: RX[%d] complete\n",
+		    nic, nic->rx_ring.sw_head);
+		atl_ring_next_dx(&nic->rx_ring.sw_head);
+	}
+}
+
+/**
+* Poll for completed and received packets
+*
+* @v netdev		Network device
+*/
+static void atl_poll(struct net_device *netdev)
+{
+	struct atl_nic *nic = netdev->priv;
+
+	/* Check link state */
+	atl_check_link(netdev);
+
+	/* Poll for TX completions, if applicable */
+	/* if (icr & ATL_IRQ_TX) */
+	atl_poll_tx(netdev);
+
+	/* Poll for RX completions, if applicable */
+	/* if (icr & ATL_IRQ_RX) */
+	atl_poll_rx(netdev);
+
+	/* Refill RX ring */
+	atl_rx_ring_fill(nic);
+}
```

Comment:
Does the hardware not provide some kind of interrupt status register that can be used to avoid unnecessarily performing all three actions (check link, poll TX, poll RX) on every poll?